src/go/k8s/internal/controller/redpanda/redpanda_controller.go

```diff
@@ -531,390 +524,294 @@ func setHelmLabelsAndAnnotations(object client.Object, rp *v1alpha2.Redpanda) {
-	var err error
-
-	// Check if HelmRelease exists or create it
-	hr := &helmv2beta2.HelmRelease{}
-
-	// have we recorded a helmRelease, if not assume we have not created it
-	if rp.Status.HelmRelease == "" {
-		// did not find helmRelease, then create it
-		hr, err = r.createHelmRelease(ctx, rp)
-		return rp, hr, err
-	}
-
-	// if we are not empty, then we assume at some point this existed, let's check
-	key := types.NamespacedName{Namespace: rp.Namespace, Name: rp.Status.GetHelmRelease()}
-	err = r.Client.Get(ctx, key, hr)
+func convertToUnstructed(resource kube.Object) (*unstructured.Unstructured, error) {
+	b, err := json.Marshal(resource)
 	if err != nil {
-		if apierrors.IsNotFound(err) {
-			rp.Status.HelmRelease = ""
-			hr, err = r.createHelmRelease(ctx, rp)
-			return rp, hr, err
-		}
-		// if this is a not found error
-		return rp, hr, fmt.Errorf("failed to get HelmRelease '%s/%s': %w", rp.Namespace, rp.Status.HelmRelease, err)
-	}
-
-	// Check if we need to update here
-	hrTemplate, errTemplated := r.createHelmReleaseFromTemplate(ctx, rp)
-	if errTemplated != nil {
-		r.event(rp, rp.Status.LastAttemptedRevision, v1alpha2.EventSeverityError, errTemplated.Error())
-		return rp, hr, errTemplated
+		return nil, err
 	}
 
-	if r.helmReleaseRequiresUpdate(ctx, hr, hrTemplate) {
-		hr.Spec = hrTemplate.Spec
-		if err = r.Client.Update(ctx, hr); err != nil {
-			r.event(rp, rp.Status.LastAttemptedRevision, v1alpha2.EventSeverityError, err.Error())
-			return rp, hr, err
-		}
-		r.event(rp, rp.Status.LastAttemptedRevision, v1alpha2.EventSeverityInfo, fmt.Sprintf("HelmRelease '%s/%s' updated", rp.Namespace, rp.GetHelmReleaseName()))
-		rp.Status.HelmRelease = rp.GetHelmReleaseName()
+	convertedObj := &unstructured.Unstructured{}
+	err = json.Unmarshal(b, convertedObj)
+	if err != nil {
+		return nil, err
 	}
 
-	return rp, hr, nil
+	return convertedObj, nil
 }
 
-func (r *RedpandaReconciler) reconcileHelmRepository(ctx context.Context, rp *v1alpha2.Redpanda) (*v1alpha2.Redpanda, *sourcev1.HelmRepository, error) {
-	// Check if HelmRepository exists or create it
-	repo := &sourcev1.HelmRepository{}
-	if err := r.Client.Get(ctx, types.NamespacedName{Namespace: rp.Namespace, Name: rp.GetHelmRepositoryName()}, repo); err != nil {
-		if apierrors.IsNotFound(err) {
-			repo = r.createHelmRepositoryFromTemplate(rp)
-			if errCreate := r.Client.Create(ctx, repo); errCreate != nil {
-				r.event(rp, rp.Status.LastAttemptedRevision, v1alpha2.EventSeverityError, fmt.Sprintf("error creating HelmRepository: %s", errCreate))
-				return rp, repo, fmt.Errorf("error creating HelmRepository: %w", errCreate)
+// based on https://github.com/fluxcd/pkg/blob/30c101fc7c9fac4d84937ff4890a3da46a9db2dd/ssa/manager_apply.go#L94-L140
+func (r *RedpandaReconciler) ServerSideApply(ctx context.Context, resource *unstructured.Unstructured) error {
```

Comment:
Rather than implement this on the controller, could we add this to the client in some way? (minus the immutable check). @andrewstucki had this implemented with a patch helper or we could attach it to the `kube.Ctl` from the helm charts?

Reply:
The problem I had with patch helper is that it uses a generator that I'm not sure is supported. 

I'm fine changing my approach to server side apply and add it to `kube.Ctl`.

REF:
Generator will produce https://github.com/redpanda-data/redpanda-operator/tree/main/src/go/k8s/api/applyconfiguration/redpanda/v1alpha2

The marker https://github.com/redpanda-data/redpanda-operator/blob/06af93fe9da36001d8ded3b82fa24fe9977d0219/src/go/k8s/api/redpanda/v1alpha2/user_types.go#L27

The applyconfiguration-gen call https://github.com/redpanda-data/redpanda-operator/blob/06af93fe9da36001d8ded3b82fa24fe9977d0219/taskfiles/k8s.yml#L32-L37

Reply:
@chrisseto that might be slightly hard with how we're returning the full `client.Object` that needs to be created. Given that I'm thinking rather than using `applyconfiguration` we might be able just to directly use `client.Patch` with `client.Apply` here as we're purely doing a raw marshaling of our objects to JSON and shouldn't have to worry about a minimal diff containing odd zero-values that weren't intended.